asrtoolkit/alignment/align_utils.py

```diff
@@ -0,0 +1,491 @@
+
+        for i in range(len(matched_segments[:-1])):
+            m1, m2 = matched_segments[i]
+            m1_next, m2_next = matched_segments[i + 1]
+            slot1 = hyp_segment[m1.end + 1:m1_next.start]
+            slot2 = ref_segment[m2.end + 1:m2_next.end]
+            l = [slot1, slot2]
+            LOGGER.debug(i, ":", l)
+
+            if len(slot1) > 1 and len(slot2) > 1:
+                unmatched_segments.append(l)
+
+        # special cases
+        # --> segment btw interval start and first match & last match and interval end
+        hyp_start, hyp_end, ref_start, ref_end = self.start_and_end_token(
+            hyp=hyp_segment, ref=ref_segment)
+
+        m1_first, m2_first = matched_segments[0]
+        if m1_first.start != hyp_start and m2_first.start != ref_start:
+            slot1_first = hyp_segment[:m1_first.start]
+            slot2_first = ref_segment[:m2_first.start]
+            l = [slot1_first, slot2_first]
+
+            if len(slot1_first) > 1 and len(slot2_first) > 1:
+                unmatched_segments.insert(0, l)
+
+        m1_last, m2_last = matched_segments[-1]
+        if m1_last.end != hyp_end and m2_last.end != ref_end:
+            slot1_last = hyp_segment[m1_last.end + 1:]
+            slot2_last = ref_segment[m2_last.end + 1:]
+
+            if len(slot1_last) > 1 and len(slot2_last) > 1:
+                unmatched_segments.append(l)
+
+        return unmatched_segments
+
+    def find_alignments(self, extract_id):
+        all_matched_segments, all_unmatched_segments = [], []
+
+        for interval_id, interval in enumerate(self.unmatched_intervals):
+            matched_segments = self.find_matched_intervals(
+                interval=interval, extractor=self.extractors[extract_id])
+
+            if matched_segments:
+                all_matched_segments.extend(matched_segments)
+
+                # find_unmatched_intervals(self, interval, matched_intervals)
+                unmatched_segments = self.find_unmatched_segments(
+                    interval=interval, matched_segments=matched_segments)
+
+                try:
+                    for segment in unmatched_segments:
+                        all_unmatched_segments.append(segment)
+                except TypeError:  # NoneType object not iterable
+                    LOGGER.debug(interval_id, "->", unmatched_segments)
+
+            else:
+
+                all_unmatched_segments.extend(
+                    interval if hasattr(interval, "__iter__") else [interval])
+
+                try:
+                    assert_sorted(all_matched_segments)
+                    assert_sorted(
+                        all_unmatched_segments)  # got token -> typeError
+                except (TypeError, AssertionError):
+                    LOGGER.debug("sorting type error - interval_id {}".format(
+                        interval_id))
+                    pass
+
+        if not all_matched_segments:
+            pass
```

Comment:
Would it be useful to add a logging statement here so we don't pass through this code without information?

Reply:
changed logic here